File: src/main/java/cc/polyfrost/vanillahud/hud/BossBar.java
package cc.polyfrost.vanillahud.hud;

import cc.polyfrost.oneconfig.config.Config;
import cc.polyfrost.oneconfig.config.annotations.*;
import cc.polyfrost.oneconfig.config.core.OneColor;
import cc.polyfrost.oneconfig.config.data.*;
import cc.polyfrost.oneconfig.events.EventManager;
import cc.polyfrost.oneconfig.hud.SingleTextHud;
import cc.polyfrost.oneconfig.libs.universal.*;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.*;
import net.minecraft.entity.boss.BossStatus;

public class BossBar extends Config {

    @HUD(
            name = "Boss Bar"
    )
    public BossBarHUD hud = new BossBarHUD();

    public BossBar() {
        super(new Mod("Boss Bar", ModType.HUD, "/vanillahud_dark.svg"), "bossbar.json");
        initialize();
    }

    public static class BossBarHUD extends SingleTextHud {

        @Switch(
                name = "Render Text"
        )
        public boolean renderText = true;

        @Switch(
                name = "Render Health"
        )
        public boolean renderHealth = true;

        @Slider(
                name = "Bar Position",
                min = 0,
                max = 100
        )
        public float barPosition = 50;

        /** Gets OneConfig's Universal Minecraft instance. */
        @Exclude
        public static final Minecraft mc = UMinecraft.getMinecraft();

        /** Gets OneConfig's Universal Minecraft fontRenderer. */
        @Exclude public static final FontRenderer fontRenderer = UMinecraft.getFontRenderer();

        /**
         * The Boss Bar width
         */
        @Exclude public static final int BAR_WIDTH = 182;

        public BossBarHUD() {
            super("", true, 1920f / 2, 2f, 1, false, false, 0, 0, 0, new OneColor(0, 0, 0, 120), false, 2, new OneColor(0, 0, 0));
            this.textType = 1;
            EventManager.INSTANCE.register(this);
        }

        @Override
        protected String getText(boolean example) {
            return this.getTextFrequent(example);
        }

        protected String getTextFrequent(boolean example) {
            return this.isBossActive() ? BossStatus.bossName : "Wither";
        }

        @Override
        public void draw(UMatrixStack matrices, float x, float y, float scale, boolean example) {
            UGraphics.GL.pushMatrix();
            UGraphics.GL.scale(scale, scale, 1);
            UGraphics.GL.translate(x / scale, y / scale, 1);
            this.drawHealth(this.getCompleteText(this.getText(example)), this.isBossActive() ? BossStatus.healthScale : 0.8f, 0, this.renderText ? 10 : 0);
            UGraphics.GL.popMatrix();
            if (this.renderText) {
                super.draw(matrices, x + this.getWidth(1.0f, example) / 2 - (float) (fontRenderer.getStringWidth(this.getCompleteText(this.getText(example))) / 2), y, scale, example);
            }
        }

        @Override
        protected boolean shouldShow() {
            return this.isBossActive() && super.shouldShow();
        }

        private boolean isBossActive() {
            return BossStatus.bossName != null && BossStatus.statusBarTime > 0;
        }

        public void drawHealth(String bossName, float healthScale, float x, float y) {
            if (this.isBossActive()) {
                --BossStatus.statusBarTime;
            }

            UGraphics.enableBlend();
            UGraphics.color4f(1.0F, 1.0F, 1.0F, 1.0F);
            mc.getTextureManager().bindTexture(Gui.icons);

            if (this.renderText && fontRenderer.getStringWidth(bossName) > BAR_WIDTH) {
                x += (fontRenderer.getStringWidth(bossName) - BAR_WIDTH) * this.barPosition / 100.0F;
            }

            float remainingHealth = healthScale * BAR_WIDTH;
            if (this.renderHealth) {
                mc.ingameGUI.drawTexturedModalRect(x, y, 0, 74, BAR_WIDTH, 5);
                mc.ingameGUI.drawTexturedModalRect(x, y, 0, 74, BAR_WIDTH, 5);
                if (remainingHealth > 0) {
                    mc.ingameGUI.drawTexturedModalRect(x, y, 0, 79, (int) remainingHealth, 5);
                }
            }

            UGraphics.disableBlend();
        }

        @Override
        protected float getWidth(float scale, boolean example) {
            float textWidth = this.renderText ? UMinecraft.getFontRenderer().getStringWidth(this.getCompleteText(getText(example))) : 0.0f;
            float healthWidth = this.renderHealth ? this.BAR_WIDTH : 0.0f;
            return Math.max(textWidth, healthWidth) * scale;
        }

        @Override
        protected float getHeight(float scale, boolean example) {
            float height = 0.0f;

            if (this.renderHealth) {
                height += 5.0F;
            }

            if (this.renderText) {
                height += 9.0F;
            }

            if (this.renderText && this.renderHealth) {
                height += 1.0F;
            }

            return height * scale;
        }
    }
}

File: src/main/kotlin/cc/woverflow/vanillahud/BossBar.kt
package cc.woverflow.vanillahud

import cc.polyfrost.oneconfig.config.Config
import cc.polyfrost.oneconfig.config.annotations.Dropdown
import cc.polyfrost.oneconfig.config.annotations.HUD
import cc.polyfrost.oneconfig.config.annotations.Switch
import cc.polyfrost.oneconfig.config.data.Mod
import cc.polyfrost.oneconfig.config.data.ModType
import cc.polyfrost.oneconfig.hud.BasicHud
import cc.polyfrost.oneconfig.libs.universal.UGraphics
import cc.polyfrost.oneconfig.libs.universal.UMatrixStack
import cc.polyfrost.oneconfig.libs.universal.UMinecraft
import cc.polyfrost.oneconfig.renderer.RenderManager
import net.minecraft.entity.boss.BossStatus
import kotlin.math.max

object BossBar : Config(Mod("Boss Bar", ModType.HUD), "bossbar.json") {
    @HUD(
        name = "HUD"
    )
    var hud = BossBarHud()

    class BossBarHud : BasicHud(true, 1920f / 2 - 5f, 2f) {
        @Dropdown(name = "Text Type", options = ["No Shadow", "Shadow", "Full Shadow"])
        var textType = 0

        @Switch(
            name = "Enable Text",
            category = "Bossbar"
        )
        var bossBarText = true

        @Switch(
            name = "Enable Health Bar",
            category = "Bossbar"
        )
        var bossBarBar = true
        override fun draw(matrices: UMatrixStack?, x: Float, y: Float, scale: Float, example: Boolean) {
            --BossStatus.statusBarTime
            if (bossBarBar) {
                val width = (182f * scale).toInt()
                val health = (BossStatus.healthScale * (width + 1)).toInt()
                UMinecraft.getMinecraft().ingameGUI.drawTexturedModalRect(x, y, 0, 74, width, 5)
                UMinecraft.getMinecraft().ingameGUI.drawTexturedModalRect(x, y, 0, 74, width, 5)
                if (health > 0) {
                    UMinecraft.getMinecraft().ingameGUI.drawTexturedModalRect(x, y, 0, 79, health, 5)
                }
            }

            if (bossBarText) {
                RenderManager.drawScaledString(BossStatus.bossName,
                    x,
                    y,
                    16777215, RenderManager.TextType.toType(textType), scale)
            }
        }

        override fun getWidth(scale: Float, example: Boolean): Float {
            if (UMinecraft.getMinecraft().ingameGUI == null) return 10f * scale
            return max(UMinecraft.getFontRenderer().getStringWidth(BossStatus.bossName).toFloat() * scale, 182f * scale)
        }

        override fun getHeight(scale: Float, example: Boolean): Float {
            var height = 0
            if (bossBarBar) {
                height += 15
            }
            if (bossBarText) {
                height += 10
            }
            return height.toFloat()
        }

        override fun shouldShow(): Boolean {
            if (!super.shouldShow()) return false
            if (BossStatus.bossName == null || BossStatus.statusBarTime > 0) return false
            return true
        }

    }
}